Add back-compat to old WP versions for new v2.2 features

westonruter opened the pull request:
Test new (and existing) plugin features in WordPress 4.9 and above. Add todos for fixing below:

- [x] Limit AMP integration on Add Plugins screen to WordPress 5.6+.
- [x] Limit AMP integration on Add Themes screen to WordPress 5.6+.
- [x] Limit plugin activation site scan to WordPress 5.6+.
- [x] Ensure Support screen works in 5.2+ and that it is hidden in older versions.
- [x] Add `status=active` parameter to `/wp/v2/themes` REST API calls for compat with WP<5.7. (See https://core.trac.wordpress.org/ticket/50152)
- [x] Limit the `_fields` requested form the `plugins` and `themes` endpoints to be just those which are needed.
- [x] Fix Settings screen loading in WP 4.9+.
- [x] Fix Onboarding Wizard loading in WP 4.9+.

Fixes #6775.
Closes #6771.

westonruter committed:
Remove obsolete SavePostValidationEvent

Closes #6771

westonruter committed:
Check dependency_support in is_needed for AmpPlugins, AmpThemes, and PluginActivationSiteScan

westonruter committed:
Add status=active parameter to /wp/v2/themes REST API call for WP<5.7 compat

westonruter commented:
> * Add `status=active` parameter to `/wp/v2/themes` REST API calls for compat with WP<5.7. (See https://core.trac.wordpress.org/ticket/50152)

@delawski Is there a need to include non-active themes in the response? I suppose so since it seems that a parent theme is not included in the list if you add `?status=active`. If so, 949eeb3 may need to be reverted and Site Scanning be made a WP 5.7+ required feature.

westonruter commented:
It may be worthwhile to limit the queried themes and plugins to just those which are active, as if there are _many_ themes and plugins installed will they all be included in the response? Actually, that brings to mind: what if there are hundreds of installed plugins and hundreds of installed themes? It seems that there is no pagination for either of the `themes` or `plugins` endpoints, so there should be a `_fields` parameter added to limit the fields to just the ones we need. In particular, I believe we should exclude fields like `description`.

delawski committed:
Do not search for blocks when on WordPress 4.9 and older

delawski commented on includes/validation/class-amp-validation-manager.php:
```diff
@@ -666,7 +666,16 @@ public static function add_validation_error_sourcing() {
+
+		// The `WP_Block_Type_Registry` class was added in WordPress 5.0.0. Because of that it sometimes caused issues
+		// on the AMP Validated URL screen when on WordPress 4.9.
+		if ( class_exists( 'WP_Block_Type_Registry' ) ) {
```
When on WordPress 4.9, I got the following error on AMP Validated URL screen:

<img width="1504" alt="Screenshot 2021-12-08 at 10 05 50" src="https://user-images.githubusercontent.com/478735/145181575-27fce6a0-9cd1-4c92-8f39-5ae9e4f58b2d.png">

This change prevents the error to occur.

westonruter replied:
Good catch. The `add_block_source_comments` method calls `handle_block_source_comment_replacement` which then calls `WP_Block_Type_Registry::get_instance()`. So by adding this condition at the top-level, no error will occur.

delawski committed:
Fail silently if plugins/themes endpoints are not available

On WordPress 5.4 and older, the themes and plugins REST endpoints are not available. We use them in Site Scan for providing a source name, version, and author instead of a bare slug. Since this information is supplementary, we can fail silently if the endpoints are not available.

delawski committed:
Ensure Onboarding Wizard works on WordPress 4.9

delawski committed:
Fetch used plugin and theme fields only

delawski committed:
Update message copy

delawski commented:
I've pushed a few commits that should fix the Site Scan on AMP Settings screen and in the Onboarding Wizard.

The scanner works now even though the`/wp/v2/plugins` and `/wp/v2/themes` REST paths are missing. We need the themes and plugins data just to provide additional information to the user. Even without it, we still have a slug which should be useful to a user:

<img width="1106" alt="Screenshot 2021-12-08 at 11 18 27" src="https://user-images.githubusercontent.com/478735/145191188-26f71068-0240-4971-8a6a-d6b953f0d021.png">

Based on this assumption, I made the themes and plugins requests to fail silently in f57f25c. 

Also, in 148d397 I've limited the fields we're requesting for themes and plugins. I haven't touched the pagination issue, though.

Lastly, with 3f3cb57 the Onboarding Wizard should work fine on WP 4.9.

dhaval-parekh commented on src/Admin/SupportScreen.php:
```diff
@@ -106,6 +106,7 @@ public static function is_needed() {
 		return (
 			is_admin()
 			&&
+			// Note: The view_site_health_checks capability was introduced in WordPress 5.2, so this will return false in older versions.
```
Even the `view_site_health_checks` capability introduces in WordPress 5.2 but that is available in the "Health Check & Troubleshooting" plugin. So it will enable the "Support" screen.

However, the Support page is working fine in WordPress 4.9

dhaval-parekh replied:
Added fix in [c8a2d41](https://github.com/ampproject/amp-wp/pull/6773/commits/c8a2d418f05ef0cd11d264ea6392cb1015f4f7c9)

westonruter replied:
When Health Check & Troubleshooting is enabled, the Support page works fine in 4.9?

dhaval-parekh replied:
The Support page is working fine in 4.9 However as of now the value of the loopback request is not coming properly and to address that we need to use a function from the "Health Check & Troubleshooting" plugin.

dhaval-parekh committed:
Add minimum WordPress version check for Support page

dhaval-parekh committed:
Update unit test cases

dhaval-parekh committed:
Fix phpcs issues

westonruter committed:
Improve theme/plugin unavailability messaging

westonruter committed:
Ensure ID attribute is added to scripts in WP<5.5

westonruter commented:
When testing Twenty Seventeen in WP 4.9, I was surprised to see a script error coming from `comment-reply.min.js`:

![image](https://user-images.githubusercontent.com/134745/145264032-71f68948-9c93-43b4-a0b3-1dbae95c2f33.png)

This is due to the `id` attribute being added to scripts starting with WP 5.5. So this is fixed with cff0d8f.

westonruter committed:
Harden injection of ID attribute into script loader tag

westonruter committed:
Fix regex pattern inversion

westonruter commented:
In Wordpress 5.2.13 I see something weird where the REST API call to obtain the settings data is:

```
https://wordpress-stable.lndo.site/amp/v1/options?_locale=user 
```

Note there is no `/wp-json/` in the URL. This is resulting in the error screen being shown:

![image](https://user-images.githubusercontent.com/134745/145287278-a05df0dc-897c-47ea-891d-55d524699dc9.png)

This is true of all the REST API URLs:

![image](https://user-images.githubusercontent.com/134745/145287478-b3949c76-efdd-4645-bd6b-63198bdf5d84.png)

This error doesn't occur in WP 5.3.10, so there's something uniquely broken about WP 5.2.x.

westonruter commented:
~I just noticed in WordPress 4.9.18..5.3.10..? after selecting Legacy Reader mode that~ 

In Reading Settings when I configure WordPress to show a specific page for the homepage and posts page:

![image](https://user-images.githubusercontent.com/134745/145293948-37b01507-15e8-47fd-bab8-8c3d4beee781.png)

The result is at the moment _no_ homepage entry being added to the list of scannable URLs:

https://github.com/ampproject/amp-wp/blob/4cd08b2982682b5641c8376694f5f5ad414ee636/src/Validation/ScannableURLProvider.php#L110-L119

This results in the review screen of the Onboarding Wizard not including the homepage. This is a bug with the `ScannableURLProvider`. It should be including here two entries for Homepage and Blog. If I instead select to show the Latest Posts on the homepage, then it does include the Homepage entry. The issue is that upon finishing the Onboarding Wizard, the Scannable URLs are no longer being re-fetched according to the latest saved settings. This is resulting in inappropriate URLs being listed, such as for legacy Reader mode:

> ![image](https://user-images.githubusercontent.com/134745/145289104-a48754f2-e368-4be6-9c58-b5d78749bad3.png)

Filed as https://github.com/ampproject/amp-wp/pull/6777

westonruter committed:
Prevent access to validation UI when dependency_support is absent

Fixes #6775

westonruter committed:
Fix URLValidationRESTControllerTest after new DependencySupport dependency added to UserAccess

westonruter committed:
Block access to AMP Customizer if site lacks dependency_support

westonruter commented:
As noted in https://github.com/ampproject/amp-wp/issues/6776#issuecomment-989288635 regarding the Customizer failing to load due to JS errors, in 75d902cfc40fa3856df7d13778528ada03ca8570 I've added a check for when someone is attempting to go to the AMP Customizer and then show an error message instead. The AMP panel is also omitted when in Legacy Reader mode. Message the users see:

![image](https://user-images.githubusercontent.com/134745/145343241-a7f2ccee-5f39-4330-b5f3-7fc433b5a536.png)

delawski committed:
Fix typo

delawski commented:
The E2E test is passing now. There was a typo in a test case (fixed in ae04461).

westonruter committed:
Fix tests for restricting DevTools to whether DependencySupport present

westonruter commented:
> Lastly, with [3f3cb57](https://github.com/ampproject/amp-wp/commit/3f3cb575b83c9bb0d72feb26aab590894a68bc31) the Onboarding Wizard should work fine on WP 4.9.

How is it that this fixed 4.9?

westonruter committed:
Further fix PHPUnit tests after c5ef2ec0f

delawski commented:
> > Lastly, with [3f3cb57](https://github.com/ampproject/amp-wp/commit/3f3cb575b83c9bb0d72feb26aab590894a68bc31) the Onboarding Wizard should work fine on WP 4.9.
> 
> How is it that this fixed 4.9?

So my first intuition was that the Settings entry point doesn't have those 3 dependencies on the list while the Wizard has and at the same time Settings screen worked okay on 4.9 and Wizard didn't.

After diving deeper, it turns out that any dependency that is added to the switch statement in `requestToHandle` and in `requestToExternal` will be expected to be provided by WordPress Core. I.e. the `dependencies` array generated in the `amp-onboarding-wizard.asset.php` after removing those 3 entries looks like this:

```diff
- 'dependencies' => array('lodash', 'wp-api-fetch', 'wp-dom-ready', 'wp-html-entities', 'wp-i18n', 'wp-url')
+ 'dependencies' => array('lodash', 'wp-api-fetch', 'wp-i18n')
```

What it all means is that instead of expecting a dependency to be provided by WordPress Core, we're bundling it as a regular npm package. Thanks to that, we're sure that a dependency exists and is in a correct version.

While including dependencies in the output file instead of using code that may already be present on the page doesn't sound good (as we're creating a redundancy), there's not much overhead involved:

| Before<br>Without `wp-dom-ready`, `wp-html-entities`, and `wp-url` | After<br>With `wp-dom-ready`, `wp-html-entities`, and `wp-url` |
| --- | --- |
| ![Screenshot 2021-12-09 at 22 49 23](https://user-images.githubusercontent.com/478735/145482311-3b9023ca-3e68-4720-920f-eb3c6b15a67c.png) | ![Screenshot 2021-12-09 at 22 50 21](https://user-images.githubusercontent.com/478735/145482339-8a3768e1-c11f-4261-9a74-563610b01d7e.png) |

After inspecting sources of those packages, the difference of 2 KB (minified) sounds reasonable. `wp-dom-ready` and `wp-html-entities` are 0.5 KB each after minification, and `wp-url` seems to contribute for the remaining 1 KB.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/wp-back-compat

* 'develop' of github.com:ampproject/amp-wp:
  Use AMP version for wp-components style
  Add `wp-components.css` to `styles` to prevent redundant files creation
  Update webpack config for wp polyfills file generation
  Remove empty scripts after webpack compilation
  Use Reader theme in validated environment if enabled

westonruter commented:
I just tried activating the plugin build from this branch and I got a fatal error:

```
PHP Fatal error:  Uncaught AmpProject\AmpWP\Exception\InvalidService: The service ID "dependency_support" is not recognized and cannot be retrieved. in /app/public/wp-content/plugins/amp/src/Exception/InvalidService.php:57
Stack trace:
#0 /app/public/wp-content/plugins/amp/src/Infrastructure/ServiceContainer/SimpleServiceContainer.php(39): AmpProject\AmpWP\Exception\InvalidService::from_service_id('dependency_supp...')
#1 /app/public/wp-content/plugins/amp/src/Services.php(55): AmpProject\AmpWP\Infrastructure\ServiceContainer\SimpleServiceContainer->get('dependency_supp...')
#2 /app/public/wp-content/plugins/amp/src/Admin/AmpPlugins.php(65): AmpProject\AmpWP\Services::get('dependency_supp...')
#3 /app/public/wp-content/plugins/amp/src/Infrastructure/ServiceBasedPlugin.php(413): AmpProject\AmpWP\Admin\AmpPlugins::is_needed()
#4 /app/public/wp-content/plugins/amp/src/Infrastructure/ServiceBasedPlugin.php(385): AmpProject\AmpWP\Infrastructure\ServiceBasedPlugin->maybe_register_service('admin.amp_plugi...', 'AmpProject\\AmpW...')
 in /app/public/wp-content/plugins/amp/src/Exception/InvalidService.php on line 57
```

Investigating.

westonruter commented on src/Admin/AmpPlugins.php:
```diff
@@ -61,6 +62,10 @@ public static function get_registration_action() {
 	 */
 	public static function is_needed() {
 
+		if ( ! Services::get( 'dependency_support' )->has_support() ) {
+			return false;
+		}
```
The inclusion of this and the same code in `AmpThemes::is_needed()` is causing fatal errors:

```
PHP Fatal error:  Uncaught AmpProject\AmpWP\Exception\InvalidService: The service ID "dependency_support" is not recognized and cannot be retrieved. in /app/public/wp-content/plugins/amp/src/Exception/InvalidService.php:57
Stack trace:
#0 /app/public/wp-content/plugins/amp/src/Infrastructure/ServiceContainer/SimpleServiceContainer.php(39): AmpProject\AmpWP\Exception\InvalidService::from_service_id('dependency_supp...')
#1 /app/public/wp-content/plugins/amp/src/Services.php(55): AmpProject\AmpWP\Infrastructure\ServiceContainer\SimpleServiceContainer->get('dependency_supp...')
#2 /app/public/wp-content/plugins/amp/src/Admin/AmpPlugins.php(65): AmpProject\AmpWP\Services::get('dependency_supp...')
#3 /app/public/wp-content/plugins/amp/src/Infrastructure/ServiceBasedPlugin.php(413): AmpProject\AmpWP\Admin\AmpPlugins::is_needed()
#4 /app/public/wp-content/plugins/amp/src/Infrastructure/ServiceBasedPlugin.php(385): AmpProject\AmpWP\Infrastructure\ServiceBasedPlugin->maybe_register_service('admin.amp_plugi...', 'AmpProject\\AmpW...')
 in /app/public/wp-content/plugins/amp/src/Exception/InvalidService.php on line 57
```

@schlessera Any idea why this would be? I thought this would have been fixed by https://github.com/ampproject/amp-wp/pull/6484.

westonruter replied:
I can fix this by making `dependency_support` support appear first in the `SERVICES` list:

```diff
diff --git a/src/AmpWpPlugin.php b/src/AmpWpPlugin.php
index 01575d06e..197f0e82d 100644
--- a/src/AmpWpPlugin.php
+++ b/src/AmpWpPlugin.php
@@ -69,6 +69,7 @@ final class AmpWpPlugin extends ServiceBasedPlugin {
 	 * @var string[]
 	 */
 	const SERVICES = [
+		'dependency_support'                 => DependencySupport::class,
 		'admin.analytics_menu'               => Admin\AnalyticsOptionsSubmenu::class,
 		'admin.google_fonts'                 => Admin\GoogleFonts::class,
 		'admin.onboarding_menu'              => Admin\OnboardingWizardSubmenu::class,
@@ -91,7 +92,6 @@ final class AmpWpPlugin extends ServiceBasedPlugin {
 		'cli.validation_command'             => Cli\ValidationCommand::class,
 		'css_transient_cache.ajax_handler'   => Admin\ReenableCssTransientCachingAjaxAction::class,
 		'css_transient_cache.monitor'        => BackgroundTask\MonitorCssTransientCaching::class,
-		'dependency_support'                 => DependencySupport::class,
 		'dev_tools.block_sources'            => DevTools\BlockSources::class,
 		'dev_tools.callback_reflection'      => DevTools\CallbackReflection::class,
 		'dev_tools.error_page'               => DevTools\ErrorPage::class,
```

westonruter replied:
There are other services with `is_needed` methods that are calling `Services::get( 'dependency_support' )->has_support()`:

* `PluginActivationSiteScan`
* `ValidationCounts`
* `Polyfills`
* `PairedBrowsing`
* `UserAccess`

They aren't causing errors like `AmpPlugins` and `AmpThemes`, apparently, because either they are `Delayed` or because they appear alphabetically later in the list of `SERVICES` than `dependency_support`.

westonruter replied:
It's strange that this is only happening at activation and not after having been activated.

westonruter replied:
Error happens only during activation in the admin, not when activating via WP-CLI.

westonruter replied:
I've temporarily changed the order of the `SERVICES`: c1e18f8a5.

westonruter commented on src/Admin/AmpThemes.php:
```diff
@@ -60,6 +61,10 @@ public static function get_registration_action() {
 	 */
 	public static function is_needed() {
 
+		if ( ! Services::get( 'dependency_support' )->has_support() ) {
+			return false;
+		}
```
As mentioned above in https://github.com/ampproject/amp-wp/pull/6773/files#r767010989, this code is causing an error upon activation:

![image](https://user-images.githubusercontent.com/134745/145646963-c631f9ed-a817-4186-9626-7f7cb82693ec.png)

westonruter commented:
> While including dependencies in the output file instead of using code that may already be present on the page doesn't sound good (as we're creating a redundancy), there's not much overhead involved:

Humm. But `wp-html-entities`, `wp-i18n`, and `wp-url` scripts should all be getting registered via `Polyfills`:

https://github.com/ampproject/amp-wp/blob/7e67d494771cd74a65615cf3f443c96371b14473/src/Admin/Polyfills.php#L101-L116

And those script files are among those being included in the build:

> ![image](https://user-images.githubusercontent.com/134745/145651870-80e30518-6aad-4c48-b03f-5413c6c6f556.png)

So why weren't they being enqueued?

westonruter committed:
Temporarily register dependency_support as first service

westonruter commented:
@delawski I think this was actually fixed by #6776 since without that chance I get errors in those 3 scripts:

![image](https://user-images.githubusercontent.com/134745/145653612-cd695829-05dd-487b-8c71-a1cc732095fe.png)

If I revert your change in 3f3cb57 after having merged `develop` including #6776, I am no longer getting any errors and the Onboarding Wizard loads successfully in 4.9. Those three scripts are being included on the page.

westonruter committed:
Revert "Ensure Onboarding Wizard works on WordPress 4.9"

This reverts commit 3f3cb575b83c9bb0d72feb26aab590894a68bc31.

schlessera committed:
Add dependency_support as a requirement for AmpPlugins & AmpThemes

westonruter committed:
Add missing HasRequirements interface to PluginActivationSiteScan

westonruter committed:
Add codeCoverageIgnore comments

delawski commented on src/DevTools/UserAccess.php:
```diff
@@ -50,6 +63,14 @@ public function register() {
+		// shown, and its the user interface which is particularly problematic on older versions of WP due to the
+		// JavaScript dependencies.
+		if ( ! $this->dependency_support->has_support() ) {
+			return false; // @codeCoverageIgnore
```
It seems that because of this check, it's impossible to toggle developer tools on WP 4.9-5.5. Is it intended?

delawski replied:
> I think we should just prevent users access to validation screens in WP<5.6.

@westonruter Actually, in light of https://github.com/ampproject/amp-wp/issues/6775#issuecomment-989305532, do you think we should disable developer tools on WP<5.6, too? If yes, then we should hide the toggle here and also remove one of the steps in the Onboarding Wizard.

westonruter replied:
That's a good question. Developer tools access is broken down into two aspects:

* Capability check: the users should never have access to DevTools (or the validation nonce)
* Preference check: validation data should not be presented to the user.

In the case of a user being on an old WP install, the latter check is more appropriate because what we're wanting to do is to prevent displaying the validation _UI_ to the user since it won't work properly. Technically such a user could still manually go to the validation screens if they know the URLs, but that is OK since they still have the capability and we aren't blocking their access.

So I think we can keep the step in the Onboarding Wizard. It's just we won't link to the Validated URLs screen.

delawski replied:
I'm not sure if we're on the same page. Here's a recording of what I meant (notice that the response's payload always returns `amp_dev_tools_enabled: false` meaning the "Enable Developer Tools" toggle is useless):

https://user-images.githubusercontent.com/478735/145884203-46d73dda-7828-4823-90eb-1ededf0ba9b8.mp4

The same applies to the Technical Background step of the Wizard.

![Screenshot 2021-12-13 at 21 33 28](https://user-images.githubusercontent.com/478735/145884481-a1c99f0f-aee0-4205-b31d-060ad48f4d39.png)

westonruter replied:
Oh! I see what you're referring to.

westonruter replied:
I've addressed this in ff7e7b783 by hiding the toggle which appears on the settings screen as well as the toggle that appears on the user profile screen.

westonruter replied:
> also remove one of the steps in the Onboarding Wizard.

@delawski Oh, I didn't do this part. If you would open a PR to address this, I'd appreciate it. It's consider this not to be a blocker for 2.2, since it only affects WP<5.6 (which granted is about a third of all installs).

delawski replied:
Sure! This one seems to be easy: #6789.

westonruter committed:
Further harden regex in amp_ensure_id_attribute_on_script_loader_tag()

westonruter committed:
Restore wp_version in test tearDown method

westonruter committed:
Run amp_ensure_id_attribute_on_script_loader_tag at earliest priority

Also remove invaid codeCoverageIgnore tags

westonruter committed:
Fix tests in WP 4.9..5.1

westonruter committed:
Fix test_prepare_response_standard in WP<5.5

westonruter committed:
Add tests for AMP_Customizer_Design_Settings

westonruter committed:
Add tests for amp_bootstrap_plugin()

dhaval-parekh commented:
**QA Passed.**

✅ AMP Plugin/Theme is only highlighted in 5.6 and the newer versions of WordPress.

✅ Site scanning is disabled after plugin activation in the WordPress version prior to 5.6

✅ Support page is only available in 5.2 and the newer versions of WordPress.

✅ Did not find any issue on the "AMP Setting" screen and "Onboarding Wizard" in WordPress 4.9

westonruter committed:
Fix test_init_customizer to account for Gutenberg being active

dhaval-parekh reviewed (approved):
Changes in PR looks good to me.

westonruter committed:
Use remove_all_filters instead of remove_all_actions for purity sake

westonruter committed:
Hide DevTools toggles when dependency_support is absent

delawski commented:
**QA Passed**

✅ DevTools toggle is not rendered on WP<5.6

| WP 4.9 | WP 5.5 | WP 5.6 |
| --- | --- | --- |
| ![Screenshot 2021-12-15 at 11 32 42](https://user-images.githubusercontent.com/478735/146171354-3ee58e34-9b5d-4e57-8025-5c4ce18e6334.png) | ![Screenshot 2021-12-15 at 11 34 21](https://user-images.githubusercontent.com/478735/146171401-610c444a-1cbd-431e-b693-77b153b6eecc.png) | ![Screenshot 2021-12-15 at 11 36 22](https://user-images.githubusercontent.com/478735/146171416-d80c940d-eacb-43ad-bd50-4d98346cd41d.png) |
